libraries/Wire/src/Wire.cpp

```diff
@@ -380,6 +380,13 @@ void TwoWire::onReceiveService(uint8_t* inBytes, int numBytes)
   if(rxBufferIndex < rxBufferLength){
     return;
   }
+
+  allocateRxBuffer(numBytes);
```

Comment:
Wouldn't it be better to do as little as possible inside the ISR? - I.e do the RX buffer allocation in begin() when slave mode is detected.

Since the current implementation of onReceiveService() never appends data to the RX buffer (it always start writing at index 0) it should be safe to allocate the buffer in begin(). 
At least as long as I2C_TXRX_BUFFER_SIZE <= BUFFER_LENGTH (add a static_assert for safety)

Reply:
if numbytes is greater than the one allocated it need realloc

Reply:
yes, but it will never be greater.  I2C_TXRX_BUFFER_SIZE = BUFFER_LENGTH = 32  (at least for now)

Reply:
allocateRxBuffer(numBytes) can still be called in the ISR, just to make sure. 
In the normal case it will do nothing if the buffer is pre-allocated in begin()

Reply:
you're right in case of slave receive. I will check all those things when I will review #217